pkg/genericapiserver/openapi/openapi.go

```diff
@@ -0,0 +1,485 @@
+		if oldProperty.Property.Items.Ref != nil {
+			newProperty.Items = new(spec.SchemaOrArray)
+			newProperty.Items.Schema = new(spec.Schema)
+			newProperty.Items.Schema.Ref = spec.MustCreateRef("#/definitions/" + *oldProperty.Property.Items.Ref)
+		} else {
+			newProperty.Items = new(spec.SchemaOrArray)
+			newProperty.Items.Schema = new(spec.Schema)
+			newProperty.Items.Schema.Type, newProperty.Items.Schema.Format, err =
+				buildOpenAPIType(oldProperty.Property.Items.Type, oldProperty.Property.Items.Format)
+			if err != nil {
+				return spec.Schema{}, err
+			}
+		}
+	}
+	newProperty.Type, newProperty.Format, err =
+		buildOpenAPIType(oldProperty.Property.Type, oldProperty.Property.Format)
+	if err != nil {
+		return spec.Schema{}, err
+	}
+	return newProperty, nil
+}
+
+func (o *openAPI) buildPaths() (spec.Paths, error) {
+	paths := spec.Paths{
+		Paths: make(map[string]spec.PathItem),
+	}
+	pathsToIgnore := createTrie(o.config.IgnorePrefixes)
+	operationIdCount := make(map[string]int)
+	// Initialize operation ID count map used in finding duplicate operation IDs.
+	for _, service := range o.config.SwaggerConfig.WebServices {
+		if pathsToIgnore.HasPrefix(service.RootPath()) {
+			continue
+		}
+		for _, route := range service.Routes() {
+			operationIdCount[route.Operation]++
+		}
+	}
+	for _, w := range o.config.SwaggerConfig.WebServices {
+		rootPath := w.RootPath()
+		if pathsToIgnore.HasPrefix(rootPath) {
+			continue
+		}
+		commonParams, err := buildParameters(w.PathParameters())
+		if err != nil {
+			return paths, err
+		}
+		for path, routes := range GroupRoutesByPath(w.Routes()) {
+			if o.config.FixPath != nil {
+				path = o.config.FixPath(path)
+			}
+			inPathCommonParamsMap, err := findCommonParameters(routes)
+			if err != nil {
+				return paths, err
+			}
+			pathItem, exists := paths.Paths[path]
+			if exists {
+				return paths, fmt.Errorf("Duplicate webservice route has been found for path: %v", path)
+			}
+			pathItem = spec.PathItem{
+				PathItemProps: spec.PathItemProps{
+					Parameters: make([]spec.Parameter, 0),
+				},
+			}
+			// add web services's parameters as well as any parameters appears in all ops, as common parameters
+			for _, p := range commonParams {
+				pathItem.Parameters = append(pathItem.Parameters, p)
+			}
+			for _, p := range inPathCommonParamsMap {
+				pathItem.Parameters = append(pathItem.Parameters, p)
+			}
+			for _, route := range routes {
+				op, err := o.buildOperations(route, inPathCommonParamsMap)
+				if err != nil {
+					return paths, err
+				}
+				switch strings.ToUpper(route.Method) {
+				case "GET":
+					pathItem.Get = op
+				case "POST":
+					pathItem.Post = op
+				case "HEAD":
+					pathItem.Head = op
+				case "PUT":
+					pathItem.Put = op
+				case "DELETE":
+					pathItem.Delete = op
+				case "OPTIONS":
+					pathItem.Options = op
+				case "PATCH":
+					pathItem.Patch = op
+				}
+				if o.config.FixOperation != nil {
```

Comment:
buildOperations does not have the context to fix the operation. I can pass that context to it, but in my opinion, it makes the code more complex. I think it is ok to say we build the operation and the we fix it based on the context we have. Also this part of code should hopefully goes away after we fix api-installation code.